contracts/modules/Experimental/Checkpoints/WeightedVoteCheckpoint.sol

```diff
@@ -0,0 +1,188 @@
+pragma solidity ^0.5.0;
+
+import "../../Module.sol";
+import "../../Checkpoint/ICheckpoint.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+/**
+ * @title Checkpoint module for token weighted vote
+ * @notice This voting system uses public votes
+ * @notice In this module every token holder has voting right (Should be greater than zero)
+ * Tally will be calculated as per the weight (balance of the token holder)
+ */
+contract WeightedVoteCheckpoint is ICheckpoint, Module {
+    using SafeMath for uint256;
+
+    struct Ballot {
+        uint256 checkpointId;
+        uint256 totalSupply;
+        uint256 startTime;
+        uint256 endTime;
+        uint256 totalNumVotes;
+        uint256 totalProposals;
+        bool isActive;
+        mapping(uint256 => uint256) proposalToVote;
+        mapping(address => Vote) voteByAddress;
+    }
+
+    Ballot[] ballots;
+
+    struct Vote {
```

Comment:
Do we really need this at all? I think all this information can be encapsulated in the Ballot itself.

Reply:
How do you think it can be encapsulated in ballot? It is the information specific to the voter.

Reply:
It has Vote Time (not needed imo), Vote weight (not needed separately, we can just store the sum of vote weights) and vote. We can keep a mapping of address -> vote in the Ballot and eliminate the need of this struct.

Reply:
The Vote struct is used to keep track of the data with respect to the voter. So if we want to show the data `who votes whom`. this struct will be useful and the timestamp is used to understand the voting campaign impact according to the timing of votes.

Reply:
As I mentioned earlier, we can just keep address -> vote mapping in the ballot to keep track of who voted for whom. I don't see the need for this struct. As far as timestamp goes, we can just use event/transaction timestamp for analysis. This analysis will be done offchain so we don't need to store timestamp like that onchain.